### PR TITLE
[BEAM-8932][Cleanup] Move external PubsubIO hooks outside of PubsubIO.

### PR DESCRIPTION
**Pull request by dpcollins-google**
Also fix these so that they typecheck correctly without relying on java runtime duck typing.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by dpcollins-google** [comment]
+R: @boyuanzz

**Comment by aaltay** [comment]
R: @chamikaramj could you find a person to review this PR?

**Comment by boyuanzz** [comment]
@chamikaramj @ihji 
Hey Cham and Heejong, could one of you please help review this PR? Major changes are related to build Pubsub source and sink as cross language IO

**Comment by chamikaramj** [comment]
Cloud you create JIRAs and (break up PR as needed) to clarify context. Also title says cleanup but based on above comment seems like this is a new feature ?

**Comment by dpcollins-google** [comment]
This is cleanup, not a new feature.  This only involves https://github.com/apache/beam/pull/10476/commits/6fd05a736fef78e543701e5a2ba2a1fe832ed239, and moves things that exist in PubsubIO that don't need to be there outside of PubsubIO and make them type check correctly.

**Comment by chamikaramj** [comment]
Thanks. Please still create a JIRA to capture the context of the change and add it to title in the form [BEAM-JIRA#].
Please see following for more details on Beam contribution process.
https://beam.apache.org/contribute/

@ihji can you help with the review ?
cc: @chadrik

**Comment by chadrik** [comment]
LGTM.  

cc'ing @mxm  since he did the Kafka external transform support.

**Comment by dpcollins-google** [comment]
Just rebased this. It should be clearer now that this is just cleanup.

File: sdks/go/container/boot.go [hunk 1]
```diff
@@ -123,7 +123,7 @@ func main() {
 		"--options=" + options,
 	}
 	if info.GetStatusEndpoint() != nil {
-		args = append(args, "--status_endpoint=" + info.GetStatusEndpoint().GetUrl())
```
Comment (udim) [inline sdks/go/container/boot.go]: This seemse like an unrelated change. Please remove if so.

**Comment by udim** [comment]
Are there any tests for these external reads and writes?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/ExternalRead.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+  @Override
+  public Map<String, Class<? extends ExternalTransformBuilder>> knownBuilders() {
+    return ImmutableMap.of(URN, ReadBuilder.class);
+  }
+
+  /** Parameters class to expose the transform to an external SDK. */
+  public static class Configuration {
+
+    // All byte arrays are UTF-8 encoded strings
```
Comment (mxm) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/ExternalRead.java]: This comment is outdated. Please update.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/ExternalWrite.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  @Override
+  public Map<String, Class<? extends ExternalTransformBuilder>> knownBuilders() {
+    return ImmutableMap.of(URN, WriteBuilder.class);
+  }
+
+  /** Parameters class to expose the transform to an external SDK. */
+  public static class Configuration {
+
+    // All byte arrays are UTF-8 encoded strings
+    private String topic;
```
Comment (mxm) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/ExternalWrite.java]: This comment is outdated. Plead update.

**Comment by dpcollins-google** [comment]
> > Also fix these so that they typecheck correctly without relying on java runtime duck typing.
> 
> Could you explain what you changed with regards to this?

Previously, these classes were inside a generic context with a parameter type of the builder, but they would have thrown an exception due to a bad cast if they were used with any type other than byte[]. This makes it explicit that the output type of the PTransform is a byte[] and removes some unsafe casts that were only correct if the output type is a byte[].

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/ExternalRead.java (hunk 2 above)
Comment (dpcollins-google) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/ExternalRead.java]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/ExternalWrite.java (hunk 3 above)
Comment (dpcollins-google) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/ExternalWrite.java]: Done.

File: sdks/go/container/boot.go (hunk 1 above)
Comment (dpcollins-google) [reply sdks/go/container/boot.go]: This was applied by running the linter with too wide a scope. resolved.

**Comment by dpcollins-google** [comment]
Retest all please

**Comment by mxm** [comment]
Makes sense. Thanks for addressing this in the cleanup.

**Comment by chamikaramj** [comment]
In general, are changes to PubSubIO in this and other PRs you authored backwards compatible ? If change changes are not backwards compatible (if users cannot upgrade Beam without changes to their pipelines) i think this needs a deeper look since Beam guarantees backwards compatibility across minor versions. This is OK for external transforms part which is experimental but PubSubIO is used in production by many users.

**Comment by dpcollins-google** [comment]
From an api perspective, this change nor any of the other ones until #10481 should have any user visible effects.  That one is an attempt to add to the user api surface, but not remove anything.

**Comment by udim** [comment]
retest this please

**Comment by udim** [comment]
retest this please

**Comment by udim** [comment]
retest this please

**Comment by dpcollins-google** [comment]
retest this please

**Comment by chamikaramj** [comment]
retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Please fix the errors. For example,
https://scans.gradle.com/s/pv2iijqg4o27m/console-log?task=:sdks:java:io:google-cloud-platform:compileJava

**Comment by dpcollins-google** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by dpcollins-google** [comment]
I believe the only issue turned up was inclusion of raw non-versioned ImmutableMap. Could we retest this please?

**Comment by dpcollins-google** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by chamikaramj** [comment]
Retest this please

**Comment by yifanzou** [comment]
Run Java PreCommit
